### PR TITLE
Dynamic bitset and ring buffer

### PR DESCRIPTION
**Pull request by notYuriy**

### CODE REVIEW

File: src/lib/bitset.zig [hunk 1]
```diff
@@ -1,36 +1,75 @@
...
+            return num_bits;
+        }
+
+        var data = [_]u8{0} ** num_bytes;
+    };
+}
+
+const DynamicBitset = struct {
+    len: usize,
+    data: []u8,
```
Comment (N00byEdge) [inline src/lib/bitset.zig:30]: We probably don't have to store `data` as a `[]` here, but a `[*]` and just check the size on `init()`, and keep it taking a slice

File: src/lib/bitset.zig [hunk 2]
```diff
@@ -1,36 +1,75 @@
...
-  expect(bs.is_set(0));
-  bs.unset(0);
-  expect(!bs.is_set(0));
+test "dynamic bitset" {
+    var mem: [2]u8 = undefined;
+    var bs: DynamicBitset = DynamicBitset.init(4, &mem);
+    expect(!bs.is_set(0));
+    bs.set(0);
+    expect(bs.is_set(0));
+    bs.set(13);
```
Comment (N00byEdge) [inline src/lib/bitset.zig:71]: Using index 13 here with a size of 4?

File: src/lib/ringbuffer.zig [hunk 3]
```diff
@@ -0,0 +1,175 @@
+const expect = @import("std").testing.expect;
+
+fn RingBuffer(comptime T: type) type {
+    return struct {
+        buffer: []T,
+        head: usize,
+        tail: usize,
+        last_remove: bool,
```
Comment (N00byEdge) [inline src/lib/ringbuffer.zig:8]: maybe `last_action: enum {Remove, Add}`?

File: src/lib/bitset.zig [hunk 4]
```diff
@@ -1,36 +1,75 @@
...
+
+    pub fn size_needed(len: usize) usize {
+        return @divCeil(len, 8);
+    }
+
+    pub fn init(len: usize, data: []u8) DynamicBitset {
+        for (data) |*cell| {
+            cell.* = 0;
+        }
+        return .{ .len = len, .data = data };
```
Comment (N00byEdge) [inline src/lib/bitset.zig:40]: `len` should definitely be checked here even if we don't do the above, would have caught the bug in the test

**Review by N00byEdge — CHANGES_REQUESTED** [review]

File: src/lib/bitset.zig (hunk 4 above)
Comment (notYuriy) [reply src/lib/bitset.zig:40]: It won't as the bug in the test occurs on bitmap accesses, which are unchecked.

Comment (N00byEdge) [reply src/lib/bitset.zig:40]: Right, this wouldn't have caught it, I'm dumb

**Pull request merged**